bootloader: allow different names for the grub binary in different archs

alfonsosanchezbeato opened the pull request:
Currently, the grub binary name is always grubx64.efi. However, for
arm64 the usual name is Ubuntu us grubaa64.efi. Make it dependent on
architecture, so one or the other is expected in each case. This will
allow using grub on arm64 systems.

This implementation relies on using golang flags to have different constants on different archs. There are multiple ways of doing this, so I'm opening this as a draft to collect opinions. Probably the default kernel command line would need to change too.

mvo5 commented:
Thanks for working on this! 

Great to see us moving towards arm64 grub. Fwiw, I personally would simply make  shimBinary into 
```go
func shimBinary() string {
 switch arch.DpkgArchitecture() { 
  case "amd64": return "shimx64"
...
```
and the same for grubBinary. Feels easier than the build flags. But I don't have a super strong opinion, either way is fine, this just seems a bit less "magic".

alfonsosanchezbeato commented:
@mvo5 right, that is also a viable solution, dynamic instead of static. The drawback is a slightly more complex code and a slightly bigger snapd binary, as you need to register two grub-recovery, one per arch (eventually 3, for risc-V ;-)), and select dynamically the good one. But I agree it would be more explicit on what is happening. I do not have a very strong opinion either way :-)

alfonsosanchezbeato commented:
Maybe also the part that feels more magic could be compensated by proper commenting.

mvo5 commented:
> @mvo5 right, that is also a viable solution, dynamic instead of static. The drawback is a slightly more complex code and a slightly bigger snapd binary, as you need to register two grub-recovery, one per arch (eventually 3, for risc-V ;-)), and select dynamically the good one. But I agree it would be more explicit on what is happening. I do not have a very strong opinion either way :-)

Aha, indeed - the grub.cfg will be included three times with my suggestion, that is a bit of a downside. I slightly lean to your suggestion then but again, no strong opinion so let's see how the others feel about it :)

anonymouse64 reviewed (approved):
lgtm, I think this approach looks fine to me

pedronis commented:
Do we need to consider the scenario where a arm64 image is built on a x86 system? iirc image code uses bootloader code via boot

alfonsosanchezbeato commented:
> Do we need to consider the scenario where a arm64 image is built on a x86 system? iirc image code uses bootloader code via boot

That's something I forgot about, thanks for pointing out. I think you are right and if you try to make an arm64 on x86 with this PR you will get a grub recovery config pointing to the x86 binary, which is certainly undesirable (this is actually also the current behavior). To fix that, we could register the assets with names like `grub-recovery.cfg:{amd64,arm64}`, a bit like it is done with `grub-recovery.cfg:static-cmdline`, so we select depending on architecture. In this case, we will have to ship all possible grub configurations for all archs in all the binaries, which is of course a drawback.

There is also the additional problem of how the `bootloader` package knows the target arch when building an image. `bootloader.Options` already contains `PrepareImageTime`, but we would need to pass around a new option (`PrepareImageArch`?) that would be used to know the target arch.  We could extract the arch from the model and use when creating the options in `MakeBootableImage()` and related methods.

alfonsosanchezbeato commented:
I have changed the approach for this and used grub's `grub_cpu` variable to find out dynamically in the grub script which is the name of the grub binary that we chainload.

The drawback is that then we need to increment the edition for `grub-recovery.cfg`, so the asset needs to be updated when snapd is refreshed. Apparently updating recovering assets was not supported yet, so I've added that bit. However, here we hit again the problem that we have in piboot of how to know where the seed partition is mounted from the bootloader package.

bboozzoo commented on bootloader/grub.go:
```diff
@@ -343,11 +344,25 @@ func (g *grub) UpdateBootConfig() (bool, error) {
+	recovUpdated := false
+	if g.role == RoleRunMode {
+		// TODO: How to find out ubuntu-seed folder?
+		recovUpdated, err = genericUpdateBootConfigFromAssets(
```
I don't think the run mode bootloader wrapper should implicitly try to update the recovery one. This would be done by the caller, by first creating the instance of the right bootloader and then invoking UpdateBootConfig(), iow the policy for update would be implemented higher up the stack. There's a TODO about updating recovery in overlord/devicestate/handlers_bootconfig.go where this policy is implemented.

alfonsosanchezbeato replied:
Right, thanks for the hint. Anyway the PR is not bumping the edition number in the end, so I have removed this part as it is not needed for the moment.

bboozzoo commented on bootloader/grub.go:
```diff
@@ -447,16 +462,19 @@ func staticCommandLineForGrubAssetEdition(asset string, edition uint) string {
 
 var (
+	// Names of binaries depend on the target arch and are defined
+	// at grub_constants.go and grub_constants_{arch}.go
```
No strong opinion, but if we ever need to cross reference the names for other arch during the build, perhaps it would be better to have:
```
type grubAssetNames struct {
    grubBinary string
    shimBInary string
}

var grubAssets = map[string]grubAssetNames {
    "amd64": {grubBinary: "grubx64.efi", shimBinary: "bootx64.efi"},
    "arm64": {grubBinary: "grubaa64.efi", shimBinary: "bootaa64.efi"},
}
```
and later in the code use:
```
import "github.com/snapcore/snapd/arch"

func getAssetsForArch() grubAssetNames {
    return grubAssets[arch.DpkgArchitecture())
}
```
and later on we can add overrides through env if needed, eg. UBUNTU_STORE_ARCH like one can do during prepare-image.

alfonsosanchezbeato commented:
Re-pushed now with some changes in the line suggested by @bboozzoo

pedronis commented on bootloader/grub.go:
```diff
@@ -446,19 +447,39 @@ func staticCommandLineForGrubAssetEdition(asset string, edition uint) string {
+}
+
+func getBootAssetsForArch() bootAssetPath {
+	return bootAssetsForArch[arch.DpkgArchitecture()]
```
It seems with this implementation we need a remark that TrustedAssetsBootloader.TrustedAssets and *Chain  should be used only at runtime/on the device.  It seems to be the case ATM but we should double check. We should error at some level if we need this and PrepareImageTime was true.

pedronis replied:
my suggestion was to capture PrepareImageTime value on the bootloader and return an error from here if PrepareImageTime is true. That will mean turning this and the other helpers into methods of the grub struct.

alfonsosanchezbeato replied:
Understood, I've addressed this now

pedronis reviewed (commented):
looks good in principle, some considerations

alfonsosanchezbeato commented on bootloader/grub.go:
```diff
@@ -446,19 +447,39 @@ func staticCommandLineForGrubAssetEdition(asset string, edition uint) string {
-)
+// bootAssetsForArch contains the paths for assets for different
+// architectures in a map
+var bootAssetsForArch = map[string]bootAssetPath{
```
I've changed this now

pedronis commented on bootloader/grub.go:
```diff
@@ -446,19 +447,39 @@ func staticCommandLineForGrubAssetEdition(asset string, edition uint) string {
+}
+
+func getGrubBootAssetsForArch() grubBootAssetPath {
+	return grubBootAssetsForArch[arch.DpkgArchitecture()]
```
the tests will need changes I think, because they will fail if run on a mismatched architecture. You might need to mock arch.DpkgArchitecture() to test this behavior and work on different platforms.

alfonsosanchezbeato replied:
I've done some changes now so tests will pass on different arches, and also to make sure that different arches are tested even when running the tests in a single arch.

pedronis commented on bootloader/grub_test.go:
```diff
@@ -53,6 +56,15 @@ func (s *grubTestSuite) SetUpTest(c *C) {
+	// By default assume amd64 in the tests: there are specialized
+	// tests for other arches
+	s.realArch = arch.DpkgArchitecture()
+	arch.SetArchitecture("amd64")
```
you could use 

```
s.AddCleanup(archtest.MockArchitecture("amd64"))
```

here and avoid the TearDownTest need

alfonsosanchezbeato replied:
Nice one, I've changed this now

pedronis reviewed (approved):
thank you, one style suggestion. This needs re-review from@anonymouse64 or a 3rd review given the last batch of changes since Ian reviewed

bboozzoo reviewed (approved):
LGTM, thank you

anonymouse64 reviewed (approved):
lgtm, thanks for the changes
